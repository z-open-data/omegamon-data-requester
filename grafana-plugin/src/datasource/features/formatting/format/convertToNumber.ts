const numberAsStringRegex = /^(?:-?\d+(\.\d+)?|0[xX][0-9A-Fa-f]+)$/;

/**
 * Either returns number or throws error, if conversion fails
 */
export function convertToNumber(value: number | string): number {
  if (typeof value === 'number') {
    return value;
  }
  const trimmed = value.trim();

  const error = new Error(`Failed to convert "${trimmed}" to number`);

  if (!numberAsStringRegex.test(trimmed)) {
    throw error;
  }

  const isHex = trimmed.startsWith('0x') || trimmed.startsWith('0X');
  const parsed = isHex ? parseInt(trimmed, 16) : Number(trimmed);
  if (isNaN(parsed)) {
    throw error;
  }
  return parsed;
}
