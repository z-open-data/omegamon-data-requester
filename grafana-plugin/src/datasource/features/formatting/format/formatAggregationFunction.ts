import { AggregationFuncName } from 'datasource/domain';

export const countColumnDisplayName = 'Count of group members';

export const allowedAggrFnNameToDisplayName: Record<AggregationFuncName, string> = {
  AVG: 'Average',
  COUNT: 'Count',
  MAX: 'Max',
  MIN: 'Min',
  SUM: 'Sum',
};

export function getAggregationFunctionDisplayName(aggregationFunction: AggregationFuncName): string {
  return allowedAggrFnNameToDisplayName[aggregationFunction];
}
