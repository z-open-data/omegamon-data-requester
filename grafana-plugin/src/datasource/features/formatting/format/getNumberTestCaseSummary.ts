import { ColumnMetadata } from 'public-domain';

export function getNumberTestCaseSummary(val: string | number, columnMd: ColumnMetadata): string {
  const scaleFactorSummary = columnMd.scaleFactor ? `scaleFactor=${columnMd.scaleFactor}` : '';
  const printfSummary = columnMd.printf ? `PRINTF="${columnMd.printf}"` : '';
  const valueFormattedForPrinting = typeof val === 'string' ? `"${val}"` : val;
  const valueSummary = `val=${valueFormattedForPrinting}`;
  return `${valueSummary} ${printfSummary} ${scaleFactorSummary}`;
}
