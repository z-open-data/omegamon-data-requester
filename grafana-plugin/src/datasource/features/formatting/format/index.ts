export { convertToNumber } from './convertToNumber';
export { applyAllFormatting, applyBaseFormatting, applyGrafanaFormatting } from './format';
export { formatEnumKeysInColumnMd, formatEnumKeysInTableMd } from './formatEnumKeys';
export { formatNumber } from './formatNumber';
export { formatMetricsInResponse } from './formatResponse';
export {
  countColumnDisplayName,
  allowedAggrFnNameToDisplayName,
  getAggregationFunctionDisplayName,
} from './formatAggregationFunction';
