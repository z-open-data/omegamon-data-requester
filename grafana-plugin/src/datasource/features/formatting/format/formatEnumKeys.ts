import { TableMetadata, ColumnMetadata } from 'public-domain';

import { applyBaseFormatting } from './format';

export function formatEnumKeysInTableMd(tableMd: TableMetadata): TableMetadata {
  const columnMdEntries = Object.entries(tableMd.columns);

  const columnMdEntriesWithFormattedEnumKeys = columnMdEntries.map<[string, ColumnMetadata]>(
    ([columndId, columnMd]) => {
      const columnMdWithFormattedEnumKeys = formatEnumKeysInColumnMd(columnMd);
      return [columndId, columnMdWithFormattedEnumKeys];
    }
  );

  const columnMdMapWithFormattedEnumKeys = Object.fromEntries(columnMdEntriesWithFormattedEnumKeys);

  const tableMdWithFormattedEnumKeys: TableMetadata = {
    ...tableMd,
    columns: columnMdMapWithFormattedEnumKeys,
  };

  return tableMdWithFormattedEnumKeys;
}

export function formatEnumKeysInColumnMd(columnMd: ColumnMetadata): ColumnMetadata {
  if (!columnMd.enums) {
    return columnMd;
  }

  if (columnMd.type === 'itmTimestamp') {
    /**
     * See features/format/README.md
     */
    const { enums, ...otherProperties } = columnMd;
    const columnMdWithoutEnums: ColumnMetadata = { ...otherProperties };
    return columnMdWithoutEnums;
  }

  const enumEntries = Object.entries(columnMd.enums);

  type EnumEntry = [string, string | number];

  const enumEntriesWithFormattedKeys = enumEntries
    .map<EnumEntry | null>(([key, val]) => {
      const formattedKey = formatEnumKey(key, columnMd);

      if (formattedKey === null) {
        return null;
      }

      return [`${formattedKey}`, val];
    })
    .filter((entryOrNull): entryOrNull is EnumEntry => entryOrNull != null);

  const enumMapWithFormattedKeys = Object.fromEntries(enumEntriesWithFormattedKeys);

  const columnMdWithFormattedEnumKeys: ColumnMetadata = {
    ...columnMd,
    enums: enumMapWithFormattedKeys,
  };

  return columnMdWithFormattedEnumKeys;
}

function formatEnumKey(key: string, columnMd: ColumnMetadata): string | null {
  try {
    const formattedKey = applyBaseFormatting(key, columnMd);
    return `${formattedKey}`;
  } catch (e) {
    /**
     * Formatting can throw error sometimes
     * e.g. REALSTOR.STORTYPE column has enum with key '001A', which fails to be converted to number.
     */
    return null;
  }
}
