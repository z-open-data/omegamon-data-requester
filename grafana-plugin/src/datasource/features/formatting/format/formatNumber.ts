import { ColumnMetadata } from 'public-domain';

import { convertToNumber } from './convertToNumber';

function scale(value: number, scaleFactor: number | undefined): number {
  if (scaleFactor == null) {
    return value;
  }
  const formattedValue = value / 10 ** scaleFactor;
  return formattedValue;
}

/**
 * Formats number according to PRINTF and scaleFactor in supplied column metadata
 * @param value - decimal number or hex number (e.g. `10` or `"0XA"` or `"A"`)
 * @returns decimal number (as number) or hex number (as string)
 */
export function formatNumber(value: number | string, columnMd: ColumnMetadata): string | number {
  const asNumber = convertToNumber(value);
  return scale(asNumber, columnMd.scaleFactor);
}
