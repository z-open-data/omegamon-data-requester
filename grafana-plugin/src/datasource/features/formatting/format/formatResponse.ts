import { Field, DataQueryResponse, DataFrame, FieldType } from '@grafana/data';
import { throwIfNullish } from 'public-common';
import { ColumnMetadata, TableMetadata, WRITETIME_COLUMN_ID, writetimeColumnMetadata } from 'public-domain';

import { extractColumnIdFromFieldName } from 'datasource/domain';

import { applyBaseFormatting } from './format';
import { addColumnMetadataToFieldConfig } from './grafanaFormattingUtils';
import { isTimeColumn } from './isTimeColumn';

function formatField(fieldToMutate: Field, columnMd: ColumnMetadata | null): void {
  addColumnMetadataToFieldConfig(columnMd, fieldToMutate);
  if (columnMd != null) {
    const notEmptyValue = fieldToMutate.values.find((val) => val !== null && val !== undefined);
    if (notEmptyValue && isTimeColumn(columnMd.type, notEmptyValue)) {
      fieldToMutate.type = FieldType.time;
    }

    fieldToMutate.values = fieldToMutate.values.map((val) => {
      if (fieldToMutate.type === FieldType.time) {
        return new Date(val).getTime();
      }
      return applyBaseFormatting(val, columnMd);
    });
  }
}

/**
 * CAUTION - bad code practice ahead!
 * Mutating existing response, cause in this case it should:
 * 1) probably cause no problems
 * 2) help with performance
 * 3) help reduce code complexity
 */
export function formatMetricsInResponse(responseToMutate: DataQueryResponse, tableMds: TableMetadata[]): void {
  (responseToMutate.data as DataFrame[]).forEach((frameToMutate, idx) => {
    if (frameToMutate.name === 'situations') {
      return;
    }
    const tableMd = tableMds.find((tMd) => tMd.id === frameToMutate.name);
    throwIfNullish(tableMd, `Not found table metadata for table id '${frameToMutate.name}'`);

    frameToMutate.fields.forEach((fieldToMutate) => {
      const columnId = extractColumnIdFromFieldName(fieldToMutate.name);
      const columnMd = columnId === WRITETIME_COLUMN_ID ? writetimeColumnMetadata : tableMd.columns[columnId];
      formatField(fieldToMutate, columnMd ?? null);
    });
  });
}
