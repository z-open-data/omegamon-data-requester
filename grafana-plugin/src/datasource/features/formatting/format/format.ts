import { Field, createTheme, getDisplayProcessor } from '@grafana/data';
import { config } from '@grafana/runtime';
import { ColumnMetadata } from 'public-domain';

import { formatNumber } from './formatNumber';
import { generateField } from './grafanaFormattingUtils';

/**
 * @param value - value of attribute, as received from ITM
 * @param columnMd - column metadata.
 */
export function applyBaseFormatting(value: string | number | null, columnMd: ColumnMetadata): string | number | null {
  if (value == null) {
    return null;
  }
  if (columnMd.type === 'number') {
    const formattedNumber = formatNumber(value, columnMd);
    return formattedNumber;
  }
  return value;
}

export function applyGrafanaFormatting(value: string | number | null, field: Field): string {
  const processor = getDisplayProcessor({ field, theme: config?.theme2 ?? createTheme() });
  const processedValue = processor(value);
  return processedValue.text;
}

export function applyAllFormatting(value: string | number | null, columnMd: ColumnMetadata): string {
  const field: Field = generateField({ values: [value], columnMd });
  const baseFormattedValue = applyBaseFormatting(value, columnMd);
  // Grafana formatting is only applied for display of these units, since we dont have reverse formatting for them.
  // Might be implemented later on OMUI5-502
  if (field.config.unit === 's' || field.config.unit === 'clocks') {
    return `${baseFormattedValue}`;
  }
  return applyGrafanaFormatting(baseFormattedValue, field);
}
