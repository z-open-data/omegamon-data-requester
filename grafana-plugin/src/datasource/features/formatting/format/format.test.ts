import { EnumMap, ColumnMetadata } from 'public-domain';

import { applyAllFormatting, applyBaseFormatting } from './format';
import { formatNumber } from './formatNumber';
import { getNumberTestCaseSummary } from './getNumberTestCaseSummary';
import {
  stringEnumColumnMetadataAndValues,
  stringColumnValues,
  stringColumnMetadata,
  numberColumnMds,
  numberCasesAsNumbersAndAsStrings,
  numberEnumColumnMetadataAndValues,
} from './mocks';

describe(applyAllFormatting.name, () => {
  describe('columnMetadata.enums defined', () => {
    describe('enum match found', () => {
      describe('value is string', () => {
        stringEnumColumnMetadataAndValues.values.forEach((value) => {
          test(`${value}`, () => {
            const { columnMetadata } = stringEnumColumnMetadataAndValues;
            const enums = columnMetadata.enums as EnumMap;
            expect(applyAllFormatting(value, columnMetadata)).toEqual(enums[value]);
          });
        });
      });
      describe('value is number', () => {
        numberEnumColumnMetadataAndValues.values.forEach((value) => {
          test(`${value}`, () => {
            const { columnMetadata } = numberEnumColumnMetadataAndValues;
            const enums = columnMetadata.enums as EnumMap;
            expect(applyAllFormatting(value, columnMetadata)).toEqual(enums[value]);
          });
        });
      });
    });
    describe('enum match NOT found', () => {
      test('value is string', () => {
        const value = 'someValueThatDoesNotMatchAnyKeyInEnumsMap';
        expect(applyAllFormatting(value, stringEnumColumnMetadataAndValues.columnMetadata)).toEqual(value);
      });
      test('value is number', () => {
        const value = 42069;
        expect(applyBaseFormatting(value, numberEnumColumnMetadataAndValues.columnMetadata)).toEqual(value);
        expect(applyAllFormatting(value, numberEnumColumnMetadataAndValues.columnMetadata)).toEqual(`${value}`);
      });
      test('value is number and hex PRINTF defined', () => {
        const columnMd = {
          printf: "'%08X'",
          ...numberEnumColumnMetadataAndValues.columnMetadata,
        } as ColumnMetadata;
        expect(applyAllFormatting(42069, columnMd)).toMatchInlineSnapshot(`"42069"`);
      });
      test('value is number and scaleFactor defined', () => {
        const columnMd = {
          ...numberEnumColumnMetadataAndValues.columnMetadata,
          scaleFactor: 2,
        } as ColumnMetadata;
        expect(applyAllFormatting(42069, columnMd)).toEqual('420.69');
      });
    });
  });

  describe('strings should remain unchanged', () => {
    stringColumnValues.forEach((val) => {
      test(val, () => {
        expect(applyAllFormatting(val, stringColumnMetadata)).toEqual(val);
      });
    });
  });

  describe(`number formatting handled by ${applyBaseFormatting.name}`, () => {
    numberColumnMds.forEach((columnMd) => {
      numberCasesAsNumbersAndAsStrings.forEach((val) => {
        const testCaseSummary = getNumberTestCaseSummary(val, columnMd);
        test(testCaseSummary, () => {
          expect(applyBaseFormatting(val, columnMd as ColumnMetadata)).toEqual(formatNumber(val, columnMd));
        });
      });
    });
  });

  describe(`number formatting handled by ${applyAllFormatting.name}`, () => {
    numberColumnMds.forEach((columnMd) => {
      numberCasesAsNumbersAndAsStrings.forEach((val) => {
        const testCaseSummary = getNumberTestCaseSummary(val, columnMd);
        test(testCaseSummary, () => {
          expect(applyAllFormatting(val, columnMd as ColumnMetadata)).toMatchSnapshot();
        });
      });
    });
  });
});
