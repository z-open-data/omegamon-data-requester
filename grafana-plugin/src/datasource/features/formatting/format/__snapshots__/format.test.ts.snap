// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0"   1`] = `"0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0"   2`] = `"0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0"  scaleFactor=2 1`] = `"0.00"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0" PRINTF=""%5.1f"" scaleFactor=1 1`] = `"0.0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="000123"   1`] = `"123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="000123"   2`] = `"123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="000123"  scaleFactor=2 1`] = `"1.23"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="000123" PRINTF=""%5.1f"" scaleFactor=1 1`] = `"12.3"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0X01F"   1`] = `"31"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0X01F"   2`] = `"31"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0X01F"  scaleFactor=2 1`] = `"0.31"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0X01F" PRINTF=""%5.1f"" scaleFactor=1 1`] = `"3.1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0xA"   1`] = `"10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0xA"   2`] = `"10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0xA"  scaleFactor=2 1`] = `"0.10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val="0xA" PRINTF=""%5.1f"" scaleFactor=1 1`] = `"1.0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-1   1`] = `"-1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-1   2`] = `"-1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-1  scaleFactor=2 1`] = `"-0.01"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-1 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"-0.1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-123   1`] = `"-123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-123   2`] = `"-123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-123  scaleFactor=2 1`] = `"-1.23"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=-123 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"-12.3"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=0   1`] = `"0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=0   2`] = `"0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=0  scaleFactor=2 1`] = `"0.00"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=0 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"0.0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=1   1`] = `"1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=1   2`] = `"1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=1  scaleFactor=2 1`] = `"0.01"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=1 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"0.1"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=10   1`] = `"10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=10   2`] = `"10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=10  scaleFactor=2 1`] = `"0.10"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=10 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"1.0"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=123   1`] = `"123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=123   2`] = `"123"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=123  scaleFactor=2 1`] = `"1.23"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=123 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"12.3"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=72491   1`] = `"72491"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=72491   2`] = `"72491"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=72491  scaleFactor=2 1`] = `"724.91"`;

exports[`applyAllFormatting number formatting handled by applyAllFormatting val=72491 PRINTF=""%5.1f"" scaleFactor=1 1`] = `"7249.1"`;
